src/main/kotlin/cacophony/semantic/VarUseTypeAnalysis.kt

```diff
@@ -1,15 +1,243 @@
+                useTypeAnalysis[expr]!!.add(
+                    resolvedVariables[expr.lhs]!!,
+                    VariableUseType.READ_WRITE,
+                )
+                useTypeAnalysis[expr]!!.mergeWith(useTypeAnalysis[expr.rhs]!!)
+            }
+
+            else -> visitExpression(expr.lhs)
+        }
+    }
+
+    private fun visitAssignment(expr: OperatorBinary.Assignment) {
+        visitExpression(expr.rhs)
+        when (expr.lhs) {
+            is VariableUse -> {
+                useTypeAnalysis[expr] = UseTypesForExpression.empty()
+                useTypeAnalysis[expr]!!.add(
+                    resolvedVariables[expr.lhs]!!,
+                    VariableUseType.WRITE,
+                )
+                useTypeAnalysis[expr]!!.mergeWith(useTypeAnalysis[expr.rhs]!!)
+            }
+
+            else -> visitExpression(expr.lhs)
+        }
+    }
+
+    private fun visitUnaryOperator(expr: OperatorUnary) {
+        visitExpression(expr.expression)
+        useTypeAnalysis[expr] =
+            UseTypesForExpression.merge(
+                useTypeAnalysis[expr.expression],
+            )
+    }
+
+    private fun visitReturnStatement(expr: Statement.ReturnStatement) {
+        visitExpression(expr.value)
+        useTypeAnalysis[expr] =
+            UseTypesForExpression.merge(
+                useTypeAnalysis[expr.value],
+            )
+    }
+
+    private fun visitWhileStatement(expr: Statement.WhileStatement) {
+        visitExpression(expr.testExpression)
+        visitExpression(expr.doExpression)
+        useTypeAnalysis[expr] =
+            UseTypesForExpression.merge(
+                useTypeAnalysis[expr.testExpression],
+                useTypeAnalysis[expr.doExpression],
+            )
+    }
+
+    private fun visitIfElseStatement(expr: Statement.IfElseStatement) {
+        visitExpression(expr.testExpression)
+        visitExpression(expr.doExpression)
+        expr.elseExpression?.let { visitExpression(it) }
+        useTypeAnalysis[expr] =
+            UseTypesForExpression.merge(
+                useTypeAnalysis[expr.testExpression],
+                useTypeAnalysis[expr.doExpression],
+                useTypeAnalysis[expr.elseExpression],
+            )
+    }
+
+    private fun visitFunctionCall(expr: FunctionCall) {
+        visitExpression(expr.function)
+        useTypeAnalysis[expr] =
+            UseTypesForExpression.merge(
+                useTypeAnalysis[expr.function],
+            )
+        expr.arguments.forEach { // arguments should be marked as read-used
+            visitExpression(it)
+            useTypeAnalysis[expr]!!.mergeWith(useTypeAnalysis[it]!!)
+        }
+
+        // variables used in called function:
+        val calledFunction = resolvedVariables[expr.function]
+        if (calledFunction is Definition.FunctionDeclaration) {
+            val map =
+                useTypeAnalysis[calledFunction.body]!!.getMap().filterNot {
```

Comment:
How does this work with recursion, and in particular with mutual recursion?
I'm worried that this line will throw an exception due to body not being defined.

It seems this is the place we should use `functionAnalysis`, no?
We can look at `functionAnalysis[calledFunction].variables` and check which variables are declared outside of it (i.e. where `definedIn != calledFunction` since this means that variable comes from parent we are in)